jxl/src/entropy_coding/ans.rs

```diff
@@ -2,6 +2,410 @@
+            }
+            idx += 1;
+        }
+        let Some((_, omit_pos)) = omit_data else {
+            return Err(Error::InvalidAnsHistogram);
+        };
+        if dist.get(omit_pos + 1) == Some(&13) {
+            return Err(Error::InvalidAnsHistogram);
+        }
+
+        let mut repeat_range_idx = 0usize;
+        let mut acc = 0;
+        let mut prev_dist = 0u16;
+        for (idx, code) in dist.iter_mut().enumerate() {
+            if repeat_range_idx < repeat_ranges.len()
+                && repeat_ranges[repeat_range_idx].start <= idx
+            {
+                if repeat_ranges[repeat_range_idx].end == idx {
+                    repeat_range_idx += 1;
+                } else {
+                    *code = prev_dist;
+                    acc += *code;
+                    if acc > (1 << 12) {
+                        return Err(Error::InvalidAnsHistogram);
+                    }
+                    continue;
+                }
+            }
+
+            if *code == 0 {
+                prev_dist = 0;
+                continue;
+            }
+            if idx == omit_pos {
+                prev_dist = 0;
+                continue;
+            }
+            if *code > 1 {
+                let zeros = (*code - 1) as i16;
+                let bitcount = (shift - ((12 - zeros) >> 1)).clamp(0, zeros);
+                *code = (1 << zeros) + ((br.read(bitcount as usize)? as u16) << (zeros - bitcount));
+            }
+            prev_dist = *code;
+            acc += *code;
+            if acc > (1 << 12) {
+                return Err(Error::InvalidAnsHistogram);
+            }
+        }
+        dist[omit_pos] = (1 << 12) - acc;
+
+        Ok(alphabet_size)
+    }
+
+    // log_alphabet_size: 5 + u(2)
+    pub fn decode(br: &mut BitReader, log_alpha_size: usize) -> Result<Self> {
+        #[derive(Debug)]
+        struct WorkingBucket {
+            dist: u16,
+            alias_symbol: u16,
+            alias_offset: u16,
+            alias_cutoff: u16,
+        }
+
+        debug_assert!((5..=8).contains(&log_alpha_size));
+        let table_size = (1u16 << log_alpha_size) as usize;
+        // 4 <= log_bucket_size <= 7
+        let log_bucket_size = 12 - log_alpha_size;
+        let bucket_size = 1u16 << log_bucket_size;
+
+        let mut dist = vec![0u16; table_size];
+        let alphabet_size = if br.read(1)? != 0 {
+            if br.read(1)? != 0 {
+                // binary
+                Self::decode_dist_binary(br, &mut dist)?
+            } else {
+                // unary
+                Self::decode_dist_unary(br, &mut dist)?
+            }
+        } else if br.read(1)? != 0 {
+            // evenly distributed
+            Self::decode_dist_evenly_distributed(br, &mut dist)?
+        } else {
+            // compressed distribution info
+            Self::decode_dist_complex(br, &mut dist)?
+        };
+
+        if let Some(single_sym_idx) = dist.iter().position(|&d| d == 1 << 12) {
```

Comment:
What do you think of splitting the alias table construction into a separate function?

Also, shouldn't it be table_size instead of 1<<12 here?

Reply:
I think `1 << 12` is correct, because the sum of probability is `1 << 12`.

Reply:
Uh well, it seems that probabilities sum to `table_size` in libjxl but they sum to `1 << 12` in the spec.

Reply:
Nevermind, it's a bit confusing that libjxl and jxl-oxide differ slightly... libjxl uses `range = 1 << precision_bits` and probs sum to `range`, where `precision_bits` is always 12 (`ANS_LOG_TAB_SIZE`).

Reply:
Ah, right, the table size is always the same, it's just the alphabet size that changes, I had forgotten that ^^'.
So ignore all the `MAX` from my comments on constant names :P